### ISSUE TITLE
Understanding the effect of the PPO metaparameters

### ISSUE THREAD

**Issue by jerabaul29** [tensorforce/tensorforce#485]
I think it would be useful to collect user experience / discuss / get some help about the high-level meaning / use of the PPO meta-parameters.

This may be a more general discussion than just tensorforce, so I opened an issue on the machine learning stack exchange. Contributions there welcome.

https://ai.stackexchange.com/questions/7723/tuning-of-ppo-metaparameters-a-high-level-overview-of-what-each-parameter-does

**Comment by 16Aghnar** [comment]
For example, we defined in our agent a spec dict for the memory, where the key 'type' leads to the value 'latest'. As far as I understand, a latest memory is not a memory replay (which is what we want). That's right?

**Comment by jerabaul29** [comment]
By the way, do you think it should be necessary to renormalize the reward or not? It seems to be implementation dependent and problem dependent, unclear to me from the literature.

@AlexKuhnle

**Comment by AlexKuhnle** [comment]
@16Aghnar: Type `latest` is essentially a "short-term" on-policy memory, hence the default for policy gradient agents (which are usually "required" to be on-policy). You can change the value to `replay` (see DQN agent et al) and it should still work, but is not on-policy anymore -- nonetheless not unreasonable, in my opinion. Important, however, the `update_mode` has to still be episode-based.

@jerabaul29: Normalizing the reward can be a very reasonable thing to do, particularly in situations where the absolute value of the reward is less meaningful than the changes of values between timesteps. That's at least one intuition...

**Comment by Gor-Ren** [comment]
Some helpful info from @AlexKuhnle on gitter:

> `batching_capacity` is not a particularly meaningful parameter, more for computation complexity reasons. It refers to how many timesteps are "batched" before calling `model.observe(...)` which, if done every timestep, slows down the process quite significantly (lots of calls to TensorFlow). It doesn't have an effect on performance at all (I think... it shouldn't).
> 
> `capacity` refers to how many timesteps fit into the memory. If you use `latest`, it's essentially `steps` x `max_timesteps` in case of episodes update mode and just steps in case of timesteps. Note that `max_timesteps` is a Runner or Environment parameter, not actually an RL agent parameter (there may not be a terminal at all), hence you need to explicitly specify it.
> Hope to answer to #485 in more detail at some point... and more detailed explanations for the docu would be very welcome as PR.

(I will try and contribute to documentation once I am out of dissertation crunch mode!)

**Comment by Gor-Ren** [comment]
Some good resources for anyone else looking to understand and optimise PPO agents:

[Unity RL package guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Training-PPO.md)
[Medium article ](https://medium.com/aureliantactics/ppo-hyperparameters-and-ranges-6fc2d29bccbe)

**Comment by AlexKuhnle** [comment]
Agreed! I'm currently trying to generally clean up the project (got a bit cluttered over time) and provide better documentation, any help and/or feedback welcome.

**Comment by wil3** [comment]
I now have some time to work on this. I am trying to match the attributes in ppo.json to the hyperparameters in the original PPO paper for consistency. I am looking to use the same hyperparamters and benchmark OpenAI baselines and Tensorforce. Primarily to see how using the beta distribution (which OpenAI baselines does not have) affects performance of trained agents.  All the PPO hyperparameters are already documented in OpenAI baselines so its just a matter of defining them for this library.

What do you think the best method to collectively define the properties? One suggestion I have is to define a [JSON schema ](https://json-schema.org/) and we can add the description attribute to each property to describe suggestions as @jerabaul29  pointed out and also how they map to the original paper. Since this will be version controlled it will allow anyone to do a PR to modify the description. As an added benefit you could use the spec in Agent.from_spec function to verify the users json file is valid.  Below is a first attempt of the schema.   The other option is we just have a doc as I described in my previous post. Let me know and I can do a PR to get things going.

```JSON
{
  "definitions": {},
  "$schema": "http://json-schema.org/draft-07/schema#",
  "$id": "http://example.com/root.json",
  "type": "object",
  "title": "The Root Schema",
  "required": [
    "type",
    "update_mode",
    "memory",
    "step_optimizer",
    "subsampling_fraction",
    "optimization_steps",
    "discount",
    "entropy_regularization",
    "gae_lambda",
    "likelihood_ratio_clipping",
    "baseline_mode",
    "baseline",
    "baseline_optimizer",
    "saver",
    "summarizer",
    "execution"
  ],
  "properties": {
    "type": {
      "$id": "#/properties/type",
      "type": "string",
      "title": "The Type Schema",
      "default": "",
      "examples": [
        "ppo_agent"
      ],
      "pattern": "^(.*)$"
    },
    "update_mode": {
      "$id": "#/properties/update_mode",
      "type": "object",
      "title": "The Update_mode Schema",
      "required": [
        "unit",
        "batch_size",
        "frequency"
      ],
      "properties": {
        "unit": {
          "$id": "#/properties/update_mode/properties/unit",
          "type": "string",
          "title": "The Unit Schema",
          "default": "",
          "examples": [
            "episodes"
          ],
          "pattern": "^(.*)$"
        },
        "batch_size": {
          "$id": "#/properties/update_mode/properties/batch_size",
          "type": "integer",
          "title": "The Batch_size Schema",
          "default": 0,
          "examples": [
            10
          ]
        },
        "frequency": {
          "$id": "#/properties/update_mode/properties/frequency",
          "type": "integer",
          "title": "The Frequency Schema",
          "default": 0,
          "examples": [
            10
          ]
        }
      }
    },
    "memory": {
      "$id": "#/properties/memory",
      "type": "object",
      "title": "The Memory Schema",
      "required": [
        "type",
        "include_next_states",
        "capacity"
      ],
      "properties": {
        "type": {
          "$id": "#/properties/memory/properties/type",
          "type": "string",
          "title": "The Type Schema",
          "default": "",
          "examples": [
            "latest"
          ],
          "pattern": "^(.*)$"
        },
        "include_next_states": {
          "$id": "#/properties/memory/properties/include_next_states",
          "type": "boolean",
          "title": "The Include_next_states Schema",
          "default": false,
          "examples": [
            false
          ]
        },
        "capacity": {
          "$id": "#/properties/memory/properties/capacity",
          "type": "integer",
          "title": "The Capacity Schema",
          "default": 0,
          "examples": [
            5000
          ]
        }
      }
    },
    "step_optimizer": {
      "$id": "#/properties/step_optimizer",
      "type": "object",
      "title": "The Step_optimizer Schema",
      "required": [
        "type",
        "learning_rate"
      ],
      "properties": {
        "type": {
          "$id": "#/properties/step_optimizer/properties/type",
          "type": "string",
          "title": "The Type Schema",
          "default": "",
          "examples": [
            "adam"
          ],
          "pattern": "^(.*)$"
        },
        "learning_rate": {
          "$id": "#/properties/step_optimizer/properties/learning_rate",
          "type": "number",
          "title": "The Learning_rate Schema",
          "default": 0.0,
          "examples": [
            0.001
          ]
        }
      }
    },
    "subsampling_fraction": {
      "$id": "#/properties/subsampling_fraction",
      "type": "number",
      "title": "The Subsampling_fraction Schema",
      "default": 0.0,
      "examples": [
        0.1
      ]
    },
    "optimization_steps": {
      "$id": "#/properties/optimization_steps",
      "type": "integer",
      "title": "The Optimization_steps Schema",
      "default": 0,
      "examples": [
        50
      ]
    },
    "discount": {
      "$id": "#/properties/discount",
      "type": "number",
      "title": "The Discount Schema",
      "default": 0.0,
      "examples": [
        0.99
      ]
    },
    "entropy_regularization": {
      "$id": "#/properties/entropy_regularization",
      "type": "number",
      "title": "The Entropy_regularization Schema",
      "default": 0.0,
      "examples": [
        0.01
      ]
    },
    "gae_lambda": {
      "$id": "#/properties/gae_lambda",
      "type": "null",
      "title": "The Gae_lambda Schema",
      "default": null,
      "examples": [
        null
      ]
    },
    "likelihood_ratio_clipping": {
      "$id": "#/properties/likelihood_ratio_clipping",
      "type": "number",
      "title": "The Likelihood_ratio_clipping Schema",
      "default": 0.0,
      "examples": [
        0.2
      ]
    },
    "baseline_mode": {
      "$id": "#/properties/baseline_mode",
      "type": "string",
      "title": "The Baseline_mode Schema",
      "default": "",
      "examples": [
        "states"
      ],
      "pattern": "^(.*)$"
    },
    "baseline": {
      "$id": "#/properties/baseline",
      "type": "object",
      "title": "The Baseline Schema",
      "required": [
        "type",
        "sizes"
      ],
      "properties": {
        "type": {
          "$id": "#/properties/baseline/properties/type",
          "type": "string",
          "title": "The Type Schema",
          "default": "",
          "examples": [
            "mlp"
          ],
          "pattern": "^(.*)$"
        },
        "sizes": {
          "$id": "#/properties/baseline/properties/sizes",
          "type": "array",
          "title": "The Sizes Schema",
          "items": {
            "$id": "#/properties/baseline/properties/sizes/items",
            "type": "integer",
            "title": "The Items Schema",
            "default": 0,
            "examples": [
              32,
              32
            ]
          }
        }
      }
    },
    "baseline_optimizer": {
      "$id": "#/properties/baseline_optimizer",
      "type": "object",
      "title": "The Baseline_optimizer Schema",
      "required": [
        "type",
        "optimizer",
        "num_steps"
      ],
      "properties": {
        "type": {
          "$id": "#/properties/baseline_optimizer/properties/type",
          "type": "string",
          "title": "The Type Schema",
          "default": "",
          "examples": [
            "multi_step"
          ],
          "pattern": "^(.*)$"
        },
        "optimizer": {
          "$id": "#/properties/baseline_optimizer/properties/optimizer",
          "type": "object",
          "title": "The Optimizer Schema",
          "required": [
            "type",
            "learning_rate"
          ],
          "properties": {
            "type": {
              "$id": "#/properties/baseline_optimizer/properties/optimizer/properties/type",
              "type": "string",
              "title": "The Type Schema",
              "default": "",
              "examples": [
                "adam"
              ],
              "pattern": "^(.*)$"
            },
            "learning_rate": {
              "$id": "#/properties/baseline_optimizer/properties/optimizer/properties/learning_rate",
              "type": "number",
              "title": "The Learning_rate Schema",
              "default": 0.0,
              "examples": [
                0.001
              ]
            }
          }
        },
        "num_steps": {
          "$id": "#/properties/baseline_optimizer/properties/num_steps",
          "type": "integer",
          "title": "The Num_steps Schema",
          "default": 0,
          "examples": [
            5
          ]
        }
      }
    },
    "saver": {
      "$id": "#/properties/saver",
      "type": "object",
      "title": "The Saver Schema",
      "required": [
        "directory",
        "seconds"
      ],
      "properties": {
        "directory": {
          "$id": "#/properties/saver/properties/directory",
          "type": "null",
          "title": "The Directory Schema",
          "default": null,
          "examples": [
            null
          ]
        },
        "seconds": {
          "$id": "#/properties/saver/properties/seconds",
          "type": "integer",
          "title": "The Seconds Schema",
          "default": 0,
          "examples": [
            600
          ]
        }
      }
    },
    "summarizer": {
      "$id": "#/properties/summarizer",
      "type": "object",
      "title": "The Summarizer Schema",
      "required": [
        "directory",
        "labels"
      ],
      "properties": {
        "directory": {
          "$id": "#/properties/summarizer/properties/directory",
          "type": "null",
          "title": "The Directory Schema",
          "default": null,
          "examples": [
            null
          ]
        },
        "labels": {
          "$id": "#/properties/summarizer/properties/labels",
          "type": "array",
          "title": "The Labels Schema",
          "items": {
            "$id": "#/properties/summarizer/properties/labels/items",
            "type": "string",
            "title": "The Items Schema",
            "default": "",
            "examples": [
              "graph",
              "total-loss"
            ],
            "pattern": "^(.*)$"
          }
        }
      }
    },
    "execution": {
      "$id": "#/properties/execution",
      "type": "object",
      "title": "The Execution Schema",
      "required": [
        "type",
        "session_config",
        "distributed_spec"
      ],
      "properties": {
        "type": {
          "$id": "#/properties/execution/properties/type",
          "type": "string",
          "title": "The Type Schema",
          "default": "",
          "examples": [
            "single"
          ],
          "pattern": "^(.*)$"
        },
        "session_config": {
          "$id": "#/properties/execution/properties/session_config",
          "type": "null",
          "title": "The Session_config Schema",
          "default": null,
          "examples": [
            null
          ]
        },
        "distributed_spec": {
          "$id": "#/properties/execution/properties/distributed_spec",
          "type": "null",
          "title": "The Distributed_spec Schema",
          "default": null,
          "examples": [
            null
          ]
        }
      }
    }
  }
}
```

**Comment by wil3** [comment]
Hi @AlexKuhnle I see your concern, this is a good point, having redundant documentation would end up being a nightmare to maintain. I guess the first step would to just update the in code documentation to include the hyperparameter translation.

I'm sorry I should have mentioned in my previous post I generated that sample schema from https://www.jsonschema.net/, pattern I believe is optional and we could exclude it.

**Comment by AlexKuhnle** [comment]
Code documentation would definitely be great, and let's think about how additional documentation like schemas can be consistently added and what their contribution is. Also, if you're generally keen to get more involved in this, get in touch (in Gitter, for instance)... ;-)

**Comment by jerabaul29** [comment]
What do you think the effect of learning rates in the specs dicts is? All gradient descents are adaptive so the learning rate should adapt automatically, right? Is there still a possibility that:

- the value of the learning rate has something to say for the learning taking place 'at all', ie if the initial learning rate is too high learning stalls.

- using a form of learning rate schedule would help (and if so, how to implement it)?

**Comment by AlexKuhnle** [comment]
I would have to look up how exactly it influences adaptive learning rates, but I'm pretty sure it does and that, particularly for RL where methods are often less stable, it can make the difference between learning and no learning. I also expect that learning rate schedules help, as it's generally a common feature. It will be supported in the next version, which I'm happy to push as "beta version" soon if you want to start using the feature already.

**Comment by jerabaul29** [comment]
Sounds good. I think this is very interesting to hear, because I was a bit bitten by the idea that 'it is adaptive, so I need not care'. But maybe I should always try to take it down a bit.

**Comment by AlexKuhnle** [comment]
I think what adaptiveness addresses is the problem that it may be advantageous to vary learning rate for different weights, but choosing a "base" learning rate is still an important hyperparameter.

**Comment by AlexKuhnle** [comment]
Improved
